File: src/main/java/br/com/controle/financeiro/controller/api/linkbuilder/ClientDTOResourceAssembler.java
package br.com.controle.financeiro.controller.api.linkbuilder;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;

import br.com.controle.financeiro.controller.api.ClientController;
import br.com.controle.financeiro.model.dto.ClientDTO;

import org.springframework.hateoas.EntityModel;
import org.springframework.hateoas.server.RepresentationModelAssembler;
import org.springframework.stereotype.Component;

@Component
public class ClientDTOResourceAssembler implements RepresentationModelAssembler<ClientDTO, EntityModel<ClientDTO>> {

    @Override
    public EntityModel<ClientDTO> toModel(ClientDTO entity) {
        return new EntityModel<>(entity, linkTo(methodOn(ClientController.class).oneClient(entity.getId())).withSelfRel(),
                              linkTo(methodOn(ClientController.class).allClients()).withRel("clients"));
    }

}

File: src/main/kotlin/br/com/controle/financeiro/controller/api/linkbuilder/ClientDTOResourceAssembler.kt
package br.com.controle.financeiro.controller.api.linkbuilder

import br.com.controle.financeiro.controller.api.ClientController
import br.com.controle.financeiro.model.dto.ClientDTO
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.server.RepresentationModelAssembler
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder
import org.springframework.stereotype.Component

@Component
class ClientDTOResourceAssembler : RepresentationModelAssembler<ClientDTO, EntityModel<ClientDTO>> {
    override fun toModel(entity: ClientDTO): EntityModel<ClientDTO> {
        return EntityModel.of(
                entity,
                WebMvcLinkBuilder.linkTo(
                                WebMvcLinkBuilder.methodOn(ClientController::class.java)
                                        .oneClient(entity.id!!)
                        )
                        .withSelfRel(),
                WebMvcLinkBuilder.linkTo(
                                WebMvcLinkBuilder.methodOn(ClientController::class.java)
                                        .allClients()
                        )
                        .withRel("clients")
        )
    }
}
